crates/sui-core/src/consensus_handler.rs

```diff
@@ -369,52 +377,56 @@ impl<T: ObjectStore + Send + Sync, C: CheckpointServiceNotify + Send + Sync> Exe
             .consensus_handler_processed_bytes
             .inc_by(bytes as u64);
 
-        let transactions_to_schedule = self
-            .epoch_store
-            .process_consensus_transactions_and_commit_boundary(
-                &sequenced_transactions,
-                &end_of_publish_transactions,
-                &self.checkpoint_service,
-                &self.object_store,
-            )
-            .await
-            .expect("Unrecoverable error in consensus handler");
+        let transactions_to_schedule = {
+            let (transactions_to_schedule, consensus_commit_batch) = self
+                .epoch_store
+                .process_consensus_transactions_and_commit_boundary(
+                    &sequenced_transactions,
+                    end_of_publish_transactions,
+                    &self.checkpoint_service,
+                    &self.object_store,
+                )
+                .await
+                .expect("Unrecoverable error in consensus handler");
+
+            // The last block in this function notifies about new checkpoint if needed
+            let final_checkpoint_round = consensus_commit_batch.final_checkpoint_round();
+            let final_checkpoint = match final_checkpoint_round.map(|r| r.cmp(&round)) {
+                Some(Ordering::Less) => {
+                    debug!(
+                        "Not forming checkpoint for round {} above final checkpoint round {:?}",
+                        round, final_checkpoint_round
+                    );
+                    return;
+                }
+                Some(Ordering::Equal) => true,
+                Some(Ordering::Greater) => false,
+                None => false,
+            };
+            let checkpoint = PendingCheckpoint {
+                roots: roots.into_iter().collect(),
+                details: PendingCheckpointInfo {
+                    timestamp_ms: timestamp,
+                    last_of_epoch: final_checkpoint,
+                    commit_height: round,
+                },
+            };
+
+            consensus_commit_batch
+                .commit(checkpoint)
+                .expect("Failed to commit consensus commit batch");
```

Comment:
To be honest I believe all this could just go into the `process_consensus_transactions_and_commit_boundary` method (which we might need to rename) . Since all of this need to happen atomically , processing the transactions and creating / storing /notifiying the checkpoint seem very coherent to me. I believe doing so would save you from having to handle via the  `ConsensusCommitBatch`.  Not a blocker of course, but thinking out loud here.

Reply:
The only way to do that would be to basically move all of this code into `process_consensus_transactions_and_commit_boundary`, so I don't think it wouldn't really save any complexity, just move it to a different spot.

The reason it is not as simple as you might think is that we have to process all the consensus transactions before we can determine whether it is the final checkpoint, so there really do need to be two different "phases" here.